File: 678-Valid Parenthesis String.py
class Solution(object):
    def checkValidString(self, s):
        """
        :type s: str
        :rtype: bool
        """
# https://www.youtube.com/watch?v=h9Y3i7hhCpo
# O(n)
        min_op, max_op = 0, 0
        for c in s:
            if c == '(':
                min_op += 1
            else:
                min_op -= 1
            
            if c != ')':
                max_op += 1
            else:
                max_op -= 1
            
            if max_op < 0:
                return False
            if min_op < 0:
                min_op = 0
        
        return min_op == 0


# Use two stacks
class Solution:
    def checkValidString(self, s: str) -> bool:
        left = []
        star = []
        for i, c in enumerate(s):
            if c == "(":
                left.append(i)
            elif c == "*":
                star.append(i)
            else:
                if left:
                    left.pop()
                elif star:
                    star.pop()
                else:
                    return False
        
        while left:
            if star and left.pop() < star.pop():
                continue
            return False
        
        return True


# costant space
class Solution:
    def checkValidString(self, s: str) -> bool:
        openParen = closeParen = 0 # maximum number of left parantheses and right parantheses 
        N = len(s)
        for i in range(N):
            if s[i] in ("(", "*"):
                openParen += 1
            else:
                openParen -= 1
            
            if s[N-i-1] in (")", "*"):
                closeParen += 1
            else:
                closeParen -= 1
            
            if openParen < 0 or closeParen < 0:
                return False
        
        return True

File: C++/678-Valid Parenthesis String.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int left = 0, right = 0;

        for (char c:s) {
            if (c == '(') {
                left++;
            } else if (c == ')') {
                left--;
            } else {
                left++;
            }
            if (left < 0) return false;
        }


        for (char c:string(rbegin(s), rend(s))) {
            if (c == ')') {
                right++;
            } else if (c == '(') {
                right--;
            } else {
                right++;
            }
            if (right < 0) return false;
        }

        return true;
    }
};


// one pass
class Solution {
public:
    bool checkValidString(string s) {
        int left = 0, right = 0, N = s.size();

        for (int i=0; i<N; i++) {
            if (s[i] == '(') {
                left++;
            } else if (s[i] == ')') {
                left--;
            } else {
                left++;
            }

            if (s[N-i-1] == ')') {
                right++;
            } else if (s[N-i-1] == '(') {
                right--;
            } else {
                right++;
            }           

            if (left < 0 || right < 0) return false;
        }

        return true;
    }
};